File: backend/src/utils/completionOtpService.js
const crypto = require('crypto');
const { COMPLETION_OTP_EXPIRY_MINUTES } = require('../config/payment');

const generateCompletionOtp = () =>
  crypto.randomInt(100000, 999999).toString();

const buildCompletionOtpPayload = () => {
  const code = generateCompletionOtp();
  const expiresAt = new Date(Date.now() + COMPLETION_OTP_EXPIRY_MINUTES * 60 * 1000);
  return {
    code,
    expiresAt,
    generatedAt: new Date(),
    customerVerified: false,
    providerVerified: false,
  };
};

const verifyCompletionOtpCode = (booking, otp, role) => {
  const record = booking.completionOtp;
  if (!record?.code) {
    return { valid: false, message: 'Completion OTP not generated. Provider must mark service complete.' };
  }
  if (record.expiresAt < new Date()) {
    return { valid: false, message: 'Completion OTP expired. Ask provider to regenerate.' };
  }

  const devBypass =
    process.env.DEV_OTP_BYPASS === 'true' && otp === (process.env.DEV_OTP_CODE || '123456');

  if (!devBypass && record.code !== otp) {
    return { valid: false, message: 'Invalid OTP. Confirm with your service provider.' };
  }

  if (role === 'customer') {
    record.customerVerified = true;
    record.customerVerifiedAt = new Date();
  } else if (role === 'provider') {
    record.providerVerified = true;
    record.providerVerifiedAt = new Date();
  } else {
    return { valid: false, message: 'Invalid role' };
  }

  const bothVerified = record.customerVerified && record.providerVerified;
  return {
    valid: true,
    bothVerified,
    message: bothVerified
      ? 'Both parties verified. Remaining payment is now unlocked.'
      : `${role === 'customer' ? 'Customer' : 'Provider'} verified. Waiting for ${role === 'customer' ? 'provider' : 'customer'} confirmation.`,
  };
};

module.exports = { buildCompletionOtpPayload, verifyCompletionOtpCode };

File: backend/src/main/java/com/pujasetu/service/CompletionOtpService.java
package com.pujasetu.service;

import com.pujasetu.config.AppProperties;
import com.pujasetu.model.Booking;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.Random;

@Service
@RequiredArgsConstructor
public class CompletionOtpService {

    private final AppProperties appProperties;
    private final Random random = new Random();

    public Booking.CompletionOtp buildCompletionOtpPayload() {
        String code = String.format("%06d", 100000 + random.nextInt(900000));
        Instant now = Instant.now();
        Instant expiresAt = now.plusSeconds(appProperties.getPayment().getCompletionOtpExpiryMinutes() * 60L);
        return Booking.CompletionOtp.builder()
                .code(code)
                .generatedAt(now)
                .expiresAt(expiresAt)
                .customerVerified(false)
                .providerVerified(false)
                .build();
    }

    public record VerifyResult(boolean valid, boolean bothVerified, String message) {}

    public VerifyResult verifyCompletionOtpCode(Booking booking, String otp, String role) {
        Booking.CompletionOtp record = booking.getCompletionOtp();
        if (record == null || record.getCode() == null) {
            return new VerifyResult(false, false, "Completion OTP not generated. Provider must mark service complete.");
        }
        if (record.getExpiresAt().isBefore(Instant.now())) {
            return new VerifyResult(false, false, "Completion OTP expired. Ask provider to regenerate.");
        }

        boolean devBypass = appProperties.getOtp().isDevBypass()
                && otp.equals(appProperties.getOtp().getDevCode());

        if (!devBypass && !record.getCode().equals(otp)) {
            return new VerifyResult(false, false, "Invalid OTP. Confirm with your service provider.");
        }

        if ("customer".equals(role)) {
            record.setCustomerVerified(true);
            record.setCustomerVerifiedAt(Instant.now());
        } else if ("provider".equals(role)) {
            record.setProviderVerified(true);
            record.setProviderVerifiedAt(Instant.now());
        } else {
            return new VerifyResult(false, false, "Invalid role");
        }

        boolean bothVerified = record.isCustomerVerified() && record.isProviderVerified();
        String message = bothVerified
                ? "Both parties verified. Remaining payment is now unlocked."
                : ("customer".equals(role) ? "Customer" : "Provider") + " verified. Waiting for "
                + ("customer".equals(role) ? "provider" : "customer") + " confirmation.";

        return new VerifyResult(true, bothVerified, message);
    }
}
